### PR TITLE
Support migration of data from v0 to v1

### LINKED ISSUE #1700: Migrate data from v0.38 format to v1.0
The entire data storage format and allowed backends is changing for Porter v1. We are moving from flat file storage (without query support) to mongodb. So both the backend plugins are changing AND the storage format.

🚨 **Migrations between alpha versions are not supported. Do not use the v1 alpha of Porter with production data.**

After #1684 is merged, we cannot release a prerelease of v1 until we have a migration for existing data.

* The migration is one way and does not alter your existing data.
* We need to pick new storage schema versions for our data, and decide where to encode the porter schema version vs the cnab schema version. 

One option is to store the porter schema version on the document, and then always generate the cnab version in the new ToCNAB() functions on our document types.

For this situation I think we will need two commands to move old data to the new format:

1. `porter storage copy --source myazuretables --destination myazuremongo` - Copies the data AS IS, from one storage backend to another. This gets the data out of azure tables and into a mongo database.
2. `porter storage migrate` - Executes a data migration on the data in the currently configured storage backend. Data in the old format that has been migrated is deleted during the migration, leaving a clean set of documents.

We also require a documentation page and blog post announcing EARLY before the v1 release how the data migrations will work.

### PR DESCRIPTION
**Pull request by carolynvs**
# What does this change
The porter storage migrate command now supports migrating from v0.38.* (not previous versions) to v1.0.0-beta.1 and higher.

You must keep your old PORTER_HOME directory around and then create a new PORTER_HOME that has v1 installed. The storage migrate command reads the data from the old home, and copies it into the new database for the v1 version of porter.

Data in the old porter home is not modified at all (and our interface to interact with it is a readonly subset of the original plugin protocol to avoid any oopsies).

# What issue does it fix
Closes #1700 

# Notes for the reviewer
N/A

# Checklist
- [x] Did you write tests?
- [ ] Did you write documentation? 
- [ ] Did you change porter.yaml or a storage document record? Update the corresponding schema file.
- [ ] If this is your first pull request, please add your name to the bottom of our [Contributors][contributors] list. Thank you for making Porter better! 🙇‍♀️

# Reviewer Checklist
* Comment with /azp run test-porter-release if a magefile or build script was modified
* Comment with /azp run porter-integration if it's a non-trivial PR

[contributors]: https://porter.sh/src/CONTRIBUTORS.md

### CODE REVIEW

**Comment by carolynvs** [comment]
I am still writing the user documentation but the code is ready to review

File: pkg/storage/migrations/migration.go
Comment (VinozzZ) [inline pkg/storage/migrations/migration.go:484]: Would this result in an error if some old record has been migrated? If so, should we just ignore duplicated key error?

File: pkg/storage/migrations/migration_test.go
Comment (VinozzZ) [inline pkg/storage/migrations/migration_test.go:229]:
```
	// Validate that the Credential Set as migrated correctly
```

File: pkg/storage/migrations/migration_test.go
Comment (VinozzZ) [inline pkg/storage/migrations/migration_test.go:253]: // Validate that the Parameter Set as migrated correctly

File: pkg/storage/migrations/source_schema.go
Comment (VinozzZ) [inline pkg/storage/migrations/source_schema.go:9]: Just curious, why does this schema only supports `json` format vs `SourceParameterSet`

File: tests/integration/migration_test.go
Comment (VinozzZ) [inline tests/integration/migration_test.go:16]: Should we also make sure sensitive data is correctly migrated?

File: docs/content/storage-migrate.md
Comment (VinozzZ) [inline docs/content/storage-migrate.md:91]: Do you think we need to call out that if their old data contains any sensitive data, they also need to configure a secret store?

**Review by VinozzZ — COMMENTED** [review]
I'm just done with my first round of the review. I would like to spend some time to set up a local environment to try out this new feature tomorrow!

File: cmd/porter/storage.go
Comment (carolynvs) [reply cmd/porter/storage.go:55]: The default is a bit complicated. I tried to explain it a bit more in the storage-migrate.md document but I'll include some of that context here too.

File: pkg/storage/migrations/migration.go
Comment (carolynvs) [reply pkg/storage/migrations/migration.go:484]: I'll switch to upsert to avoid the problem.

File: pkg/storage/migrations/source_schema.go
Comment (carolynvs) [reply pkg/storage/migrations/source_schema.go:9]: These source data structs are copied directly from porter v0.38. The schema document only ever supported json, while parameter sets and credential sets supported json and yaml (since we read them from the user on commands like porter install). Porter v0.38 only ever stored parameter and credential sets as json files though when using the old filesystem storage plugin.

**Comment by carolynvs** [comment]
@VinozzZ Thanks for the review! I have a few items fixed already but some need a bit more time. I'll let you know when this is ready to look at again.

File: pkg/storage/migrations/migration.go
Comment (carolynvs) [reply pkg/storage/migrations/migration.go:281]: That's a great question. I think that the secrets protocol should define that create will essentially do a Create or Update (upsert). There are other scenarios where it's possible that Porter could retry an operation that would result in create being called on the same secret key.

File: pkg/storage/migrations/migration.go
Comment (carolynvs) [reply pkg/storage/migrations/migration.go:416]: I'd rather save to the secret store first, and possibly end up with a "dangling" secret than the flip side which would mean that we saved a reference to an external secret that we later find out we couldn't persist, leaving out database in a bad state. If we go with the secret protocol saying that Create really means CreateOrUpdate (upsert) then is this (dangling secrets) still a problem?

If we want to change how the secrets protocol / sanitizer works, I think it makes sense to articular the edge cases that we don't feel are well supported now and then change it in a separate PR so that it gets the consideration it requires.

**Comment by carolynvs** [comment]
@VinozzZ I've created a new PR with some of the fixes for the problems you identified. Once that is merged I'll reset the more recent commits on this PR (since they are really unrelated to the migration). So please hold off on reviewing new commits until I've cleaned this one up.

File: pkg/storage/migrations/migration.go
Comment (VinozzZ) [reply pkg/storage/migrations/migration.go:418]: I think if the secret plugins is doing upsert, this is probably ok

File: cmd/porter/storage.go
Comment (VinozzZ) [inline cmd/porter/storage.go:55]: `and otherwise the global namespace is used.` --> `, otherwise the global namespace is used.~

File: docs/content/storage-migrate.md
Comment (VinozzZ) [inline docs/content/storage-migrate.md:42]: Is the filesystem meant to be a link to the doc  here?

File: tests/integration/testdata/migration/installation-show-sensitive-data-output.json
Comment (VinozzZ) [inline tests/integration/testdata/migration/installation-show-sensitive-data-output.json:26]: Should this be true if it's a sensitive data?

File: pkg/storage/migrations/migration.go
Comment (carolynvs) [reply pkg/storage/migrations/migration.go:418]: I'm fixing the k8s plugin to upsert 👍

File: tests/integration/testdata/migration/installation-show-sensitive-data-output.json
Comment (carolynvs) [reply tests/integration/testdata/migration/installation-show-sensitive-data-output.json:26]: The name parameter is not defined as sensitive, just the password parameter.

https://github.com/getporter/examples/blob/571103890795923800762b132ac7b22afb28ff28/sensitive-data/porter.yaml#L7-L13

File: tests/integration/testdata/migration/secrets/01G6K8CZ08T78WXTJYHR0NTYBS-name
Comment (VinozzZ) [inline tests/integration/testdata/migration/secrets/01G6K8CZ08T78WXTJYHR0NTYBS-name:1]: If the name isn't defined as sensitive, I'm a bit confused on why the secret plugin still created a secret for this parameter

Comment (carolynvs) [reply tests/integration/testdata/migration/secrets/01G6K8CZ08T78WXTJYHR0NTYBS-name:1]: There is both a parameter "name" and an output "name". The parameter is not sensitive. The output is though

https://github.com/getporter/examples/blob/571103890795923800762b132ac7b22afb28ff28/sensitive-data/porter.yaml#L18-L21

So we have one sensitive parameter, "password" and one sensitive output "name" which is why we got 2 secrets in the filesystem plugin.

Maybe the example bundle could be improved for clarity if we changed the output to be admin-password (instead of name) and have it persist the password parameter value? I think part of the problem may be that the parameter isn't sensitive but the output is and they both have the same name.

**Review by VinozzZ — APPROVED** [review]
💯 🥇

**Pull request merged**
